### PR TITLE
Support metric evaluation in the RetinaNet

### PR DESCRIPTION
**Pull request by LukeWood**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [ ] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by LukeWood** [comment]
Failure looks unrelated

**Comment by LukeWood** [comment]
> Failure looks unrelated

nevermind, it is - need to change the test for coco callback

File: keras_cv/models/object_detection/__test_utils__.py [hunk 1]
```diff
@@ -21,13 +21,21 @@ def _create_bounding_box_dataset(
 ):
     # Just about the easiest dataset you can have, all classes are 0, all boxes are
     # exactly the same.  [1, 1, 2, 2] are the coordinates in xyxy
-    xs = tf.ones((5, 256, 256, 3), dtype=tf.float32)
-    y_classes = tf.zeros((5, 10), dtype=tf.float32)
+    xs = tf.random.normal(shape=(1, 512, 512, 3), dtype=tf.float32)
+    xs = tf.tile(xs, [5, 1, 1, 1])
```
Comment (ianstenbit) [inline keras_cv/models/object_detection/__test_utils__.py:25]: Why the switch to make all 5 images identical?

File: keras_cv/models/object_detection/retina_net/retina_net.py [hunk 2]
```diff
@@ -222,34 +244,21 @@ def _forward(self, images, training=None):
...
-                box_format=self.bounding_box_format,
-                variance=BOX_VARIANCE,
-            )
         return box_pred, cls_pred
 
     def decode_predictions(self, predictions, images):
         # no-op if default decoder is used.
         box_pred, cls_pred = predictions
+        # box_pred is on "center_yxhw" format, convert to target format.
+        anchors = self.anchor_generator(images[0])
```
Comment (ianstenbit) [inline keras_cv/models/object_detection/retina_net/retina_net.py:253]: Is it safe to assume images aren't ragged wrt. size?

File: keras_cv/models/object_detection/retina_net/retina_net.py [hunk 3]
```diff
@@ -415,27 +423,50 @@ def train_step(self, data):
...
-        return self.compute_metrics(x, {}, {}, sample_weight={})
+        box_pred, cls_pred = self(x, training=False)
+        _ = self.compute_loss(x, box_pred, cls_pred, boxes, classes)
+
+        if not self._has_user_metrics:
+            return super().compute_metrics(x, {}, {}, sample_weight={})
+        y_pred = self.decode_predictions((box_pred, cls_pred), x)
+        return self.compute_metrics(x, y, y_pred, sample_weight=None)
+
+    def compute_metrics(self, x, y, y_pred, sample_weight):
```
Comment (ianstenbit) [inline keras_cv/models/object_detection/retina_net/retina_net.py:456]: Why override this?

We should be able to just pass `metrics` to `super().compile` and the default metrics implementation should handle it.

See e.g. https://colab.sandbox.google.com/drive/1lKXIucTDaQ_50ih6d1S_kYdz31lnKHTG?resourcekey=0-3n7ckwRR1393vs0T3WP2Bw#scrollTo=j-m3LsD42lOU where I added metric support to RetinaNet without this change.

File: keras_cv/models/object_detection/retina_net/retina_net.py [hunk 4]
```diff
@@ -287,15 +297,10 @@ def compile(
                 A preconfigured `FocalLoss` is provided when the string "focal" is
                 passed.
             weight_decay: a float for variable weight decay.
+            metrics: object detection metrics.
```
Comment (ianstenbit) [inline keras_cv/models/object_detection/retina_net/retina_net.py:300]: This warrants some more docstring content -- specifically, that it will break TPU compatibility and make decoding happen during training

File: keras_cv/layers/object_detection/multi_class_non_max_suppression.py [hunk 5]
```diff
@@ -57,13 +57,13 @@ def __init__(
         self.max_detections_per_class = max_detections_per_class
         self.built = True
 
-    def call(self, box_prediction, confidence_prediction):
+    def call(self, box_prediction, class_prediction):
```
Comment (ianstenbit) [inline keras_cv/layers/object_detection/multi_class_non_max_suppression.py:60]: Maybe I'm overly indoctrinated in the small CL concept, but IMO it would be better to break the NMS + LabelEncoder changes to separate PRs. Up to you.

**Review by ianstenbit — CHANGES_REQUESTED** [review]
Thanks Luke!

File: keras_cv/models/object_detection/__test_utils__.py (hunk 1 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/__test_utils__.py:25]: otherwise it is too hard to make the test consistently overfit the dataset and thus receive a good Recall score.  When you're fitting noise I believe you end up with too many competing signals.

This was before the batchnorm though, so maybe I revert and try again?

File: keras_cv/models/object_detection/retina_net/retina_net.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net.py:253]: Good catch - I used to have a check for this in call and accidentally removed it here.  I've added it back to the call method to make sure that we catch this as soon as possible (the moment train step or eval step starts)

File: keras_cv/layers/object_detection/multi_class_non_max_suppression.py (hunk 5 above)
Comment (LukeWood) [reply keras_cv/layers/object_detection/multi_class_non_max_suppression.py:60]: Lets leave this one for now, but I am happy to be more mindful of this going forwards.

File: keras_cv/models/object_detection/__test_utils__.py (hunk 1 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/__test_utils__.py:25]: Actually, this was already making 5 identical images, but before they were all ones and now they are random uniforms.  

This allows the model to actually learn features.  When convolving across an image of ones yields the sum of ones always for all feature spaces --- it doesn't allow the model to distinguish between what is a box and what is not a box.  Its actually an impossible distribution for a convolutional neural network to learn.  The old dataset was meant to be trivial but it is actually literally impossible to solve without calling all pixels boxes.

Similarly, when all of the ys are stacked on the same location it becomes impossible for the model to say "there are 9 objects on the exact same pixel". Basically these changes are necessary to test Recall, and also make the test set closer to a real OD dataset by allowing the model to learn features, and making the boxes actually have distinct locations.

File: keras_cv/models/object_detection/retina_net/retina_net.py (hunk 3 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net.py:456]: you were only able to get this to work because we used the old box format at the time:

```
        return self.compute_metrics(x, tf.concat([y["boxes"], y["classes"]], axis=-1), preds, sample_weight={})
```

When you use bounding boxes the default `compute_metrics()` implementation interprets dictionaries as a mapping of distinct outputs.

Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net.py:456]: Unfortunately with dictionary boxes this is required.   Maybe we can come up with a utility or abstraction to make doing this easier?

File: keras_cv/models/object_detection/retina_net/retina_net.py (hunk 4 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net.py:300]: Done and updated

**Comment by ianstenbit** [comment]
@LukeWood did you forget to push your changes?
I see your comments saying you updated things but I see no new commits

**Comment by LukeWood** [comment]
> @LukeWood did you forget to push your changes? I see your comments saying you updated things but I see no new commits

Sorry.  lukewood/rnmodelmetrics vs origin/rnmodelmetrics 🤦

File: keras_cv/models/object_detection/retina_net/retina_net.py [hunk 6]
```diff
@@ -297,7 +304,17 @@ def compile(
...
                 passed.
             weight_decay: a float for variable weight decay.
-            metrics: object detection metrics.
+            metrics: KerasCV object detection metrics that accept decoded
+                bounding boxes as their inputs.  Examples of this metric type are
+                `keras_cv.metrics.BoxRecall()` and
+                `keras_cv.metrics.BoxMeanAveragePrecision()`.  When `metrics` are
+                included in the call to `compile()`, the RetinaNet will perform
+                non max suppression decoding using a
+                `keras_cv.layers.MultiClassNonMaxSuppression()` layer in the
```
Comment (ianstenbit) [inline keras_cv/models/object_detection/retina_net/retina_net.py:313]: Is this guaranteed to be this specific nms layer? What if the user specified a custom one?

**Review by ianstenbit — COMMENTED** [review]
Could you create a quick test colab to verify that this works as expected on TPU?

or maybe just use whatever colab Martin has for his testing?


Also maybe let’s add a todo in the compute metrics about improving that e.g with a Util function

File: keras_cv/models/object_detection/retina_net/retina_net.py (hunk 6 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net.py]: Good point, this was a little inaccurate so I've updated it.  Good to get it right here as we will likely copy paste this to a few models.

**Comment by LukeWood** [comment]
> Could you create a quick test colab to verify that this works as expected on TPU?
> 
> or maybe just use whatever colab Martin has for his testing?
> 
> Also maybe let’s add a todo in the compute metrics about improving that e.g with a Util function

https://colab.research.google.com/drive/1DibQzykcfRc_LNB1TPW0D5bFk7r8J6BI?usp=sharing

File: keras_cv/models/object_detection/retina_net/retina_net.py [hunk 7]
```diff
@@ -416,27 +444,50 @@ def train_step(self, data):
...
         self.optimizer.apply_gradients(zip(gradients, trainable_vars))
 
-        return self.compute_metrics(x, {}, {}, sample_weight={})
+        if not self._has_user_metrics:
+            return super().compute_metrics(x, {}, {}, sample_weight={})
+
+        y_pred = self.decode_predictions((box_pred, cls_pred), x)
+        return self.compute_metrics(x, y, y_pred, sample_weight=None)
 
     def test_step(self, data):
```
[the remark this answers is not in the document]
Comment (ianstenbit) [reply keras_cv/models/object_detection/retina_net/retina_net.py:453]: (If so -- probably need a test for this also? Maybe as a follow-up PR if you'd prefer)

File: keras_cv/models/object_detection/retina_net/retina_net.py [hunk 8]
```diff
@@ -288,15 +305,23 @@ def compile(
                 A preconfigured `FocalLoss` is provided when the string "focal" is
                 passed.
             weight_decay: a float for variable weight decay.
+            metrics: KerasCV object detection metrics that accept decoded
+                bounding boxes as their inputs.  Examples of this metric type are
+                `keras_cv.metrics.BoxRecall()` and
+                `keras_cv.metrics.BoxMeanAveragePrecision()`.  When `metrics` are
+                included in the call to `compile()`, the RetinaNet will perform
+                non max suppression decoding using the
+                `RetinaNet.decode_predictions()` metrod.  By default, this method
```
Comment (ianstenbit) [inline keras_cv/models/object_detection/retina_net/retina_net.py:314]: nit: maybe just "during the forward pass"?

File: keras_cv/models/object_detection/retina_net/retina_net.py [hunk 9]
```diff
@@ -288,15 +305,23 @@ def compile(
...
+                uses a `keras_cv.layers.MultiClassNonMaxSuppression()` layer to
+                perform decoding.  This behavior can be customized by passing in a
+                `prediction_decoder` to the constructor or by modifying the
+                `prediction_decoder` attribute on the model. It should be noted
+                that the non max suppression operation does not have TPU support,
```
Comment (ianstenbit) [inline keras_cv/models/object_detection/retina_net/retina_net.py:319]: maybe say "the default non max suppression operation"

**Review by ianstenbit — CHANGES_REQUESTED** [review]
Generally LGTM.
Were you able to get the Colab to work? It still shows errors for me (filesystem issues).

Probably worth making sure Martin's testing works with this change also.

File: keras_cv/models/object_detection/retina_net/retina_net.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/models/object_detection/retina_net/retina_net.py:453]: Probably worth doing in a follow up

**Comment by LukeWood** [comment]
> Generally LGTM. Were you able to get the Colab to work? It still shows errors for me (filesystem issues).
> 
> Probably worth making sure Martin's testing works with this change also.

Let's have Martin test it - I don't have his notebook at this time.  I don't think we should block on this personally.  The delta in the  forward pass is a Python boolean that autograph will trace out.

**Review by ianstenbit — APPROVED** [review]
LGTM thanks Luke.

Let's make sure before we cut another release that Martin's TPU notebooks are all happy with this change.

**Comment by LukeWood** [comment]
> LGTM thanks Luke.
> 
> Let's make sure before we cut another release that Martin's TPU notebooks are all happy with this change.

Absolutely.  I want to have an integration test covering TPU training + metric evaluation once we have a nice structure to do so.

**Comment by LukeWood** [comment]
Update: works on TPU: https://colab.research.google.com/drive/1DibQzykcfRc_LNB1TPW0D5bFk7r8J6BI?usp=sharing
